### PR TITLE
stmhal: fix USB HID Receive

### PR DESCRIPTION
**Pull request by prusnak**
I found a bug in the USB HID Receive code that resulted in first read packet always containing no data just zeroes. All subsequent reads were OK.

This patch fixes the issue by adding Init method to HID_fops, which sets the buffer correctly even for the first packet.

Also there is a SNAK/CNAK mechanism that implements flow control in case user does not call recv method often enough (it tells host side to stop sending more data).

### CODE REVIEW

File: stmhal/usbd_hid_interface.c [hunk 1]
```diff
@@ -45,22 +45,33 @@
...
 
 /**
+  * @brief  HID_Itf_Init
+  *         Initializes the HID media low layer
+  * @param  None
+  * @retval Result of the opeartion: USBD_OK if all operations are OK else USBD_FAIL
+  */
+static int8_t HID_Itf_Init(void)
+{
+    USBD_HID_SetRxBuffer(&hUSBDDevice, buffer[current_write_buffer]);
```
Comment (dpgeorge) [inline stmhal/usbd_hid_interface.c]: To make it robust, I think this function should initialise all the static variables (current_read_buffer, etc).

**Review by dpgeorge — COMMENTED** [review]
Thanks!

I assume you tested the read-poll function?

**Comment by prusnak** [comment]
Added initialization of static variables to Init function.

File: stmhal/usbd_hid_interface.c [hunk 2]
```diff
@@ -76,9 +90,18 @@ static int8_t HID_Itf_Receive(uint8_t* Buf, uint32_t Len) {
     USBD_HID_SetRxBuffer(&hUSBDDevice, buffer[current_write_buffer]);
     USBD_HID_ReceivePacket(&hUSBDDevice);
 
+    PCD_HandleTypeDef *hpcd = hUSBDDevice.pData;
```
Comment (dpgeorge) [inline stmhal/usbd_hid_interface.c]: Please add a short comment saying what this code does.

File: stmhal/usbd_hid_interface.c [hunk 3]
```diff
@@ -106,6 +129,10 @@ int USBD_HID_Rx(uint8_t *buf, uint32_t len, uint32_t timeout) {
     memcpy(buf, buffer[current_read_buffer], last_read_len);
     current_read_buffer = !current_read_buffer;
 
+    PCD_HandleTypeDef *hpcd = hUSBDDevice.pData;
```
Comment (dpgeorge) [inline stmhal/usbd_hid_interface.c]: Add short comment here as well.

File: stmhal/usbdev/class/inc/usbd_cdc_msc_hid.h [hunk 4]
```diff
@@ -20,6 +20,12 @@
 #define CDC_OUT_EP    (0x03)
 #define CDC_CMD_EP    (0x82)
 
+// Need to define here for usbd_hid_interface.c (it needs HID_OUT_EP_WITH_CDC)
+#define HID_IN_EP_WITH_CDC (0x81)
+#define HID_OUT_EP_WITH_CDC (0x01)
+#define HID_IN_EP_WITH_MSC (0x83)
+#define HID_OUT_EP_WITH_MSC (0x03)
```
Comment (dpgeorge) [inline stmhal/usbdev/class/inc/usbd_cdc_msc_hid.h]: These shouldn't be exposed.  Instead one should use the `hid_out_ep` variable defined in usbdev/class/src/usbd_cdc_msc_hid.c.  Two options to make it work: 1) make a function in usbd_cdc_msc_hid.c to get the hid out EP (eg USBD_HID_GetOutEP); 2) add functions to usbd_cdc_msc_hid.c to set the SNAK/CNAK bits and call those functions from usbd_hid_interface.c.  Probably the latter is cleaner.

Comment (prusnak) [inline stmhal/usbdev/class/inc/usbd_cdc_msc_hid.h]: I used the approach used in usbd_cdc_interface (exposing defines via .h file), but apporach 2) makes much more sense. Will update the PR.

**Comment by prusnak** [comment]
Reworked last commit according to comments

**Comment by dpgeorge** [comment]
Great, thank you!  Rebased and merged, see 0883a7e72f7122b307798205441c96120446bf75

**Pull request closed without merging**